src/Temporalio/Worker/Tuning/WorkerTuner.cs

```diff
@@ -0,0 +1,106 @@
+        /// </summary>
+        /// <param name="targetMemoryUsage">A value between 0 and 1 that represents the target
+        /// (system) memory usage. It's not recommended to set this higher than 0.8, since how much
+        /// memory a workflow may use is not predictable, and you don't want to encounter OOM
+        /// errors.</param>
+        /// <param name="targetCpuUsage">A value between 0 and 1 that represents the target (system)
+        /// CPU usage. This can be set to 1.0 if desired, but it's recommended to leave some
+        /// headroom for other processes.</param>
+        /// <param name="workflowOptions">Options for the workflow task slot supplier.</param>
+        /// <param name="activityOptions">Options for the activity task slot supplier.</param>
+        /// <param name="localActivityOptions">Options for the local activity slot supplier.</param>
+        /// <returns>The tuner.</returns>
+        /// <remarks>
+        /// WARNING: Resource based tuning is currently experimental.
+        /// </remarks>
+        public static WorkerTuner CreateResourceBased(
+            double targetMemoryUsage,
+            double targetCpuUsage,
+            ResourceBasedSlotSupplierOptions? workflowOptions = null,
+            ResourceBasedSlotSupplierOptions? activityOptions = null,
+            ResourceBasedSlotSupplierOptions? localActivityOptions = null)
+        {
+            ResourceBasedTunerOptions tunerOpts =
+                new ResourceBasedTunerOptions(targetMemoryUsage, targetCpuUsage);
+            return new CompositeTuner(
+                new ResourceBasedSlotSupplier(
+                    workflowOptions == null
+                        ? new ResourceBasedSlotSupplierOptions()
+                        : workflowOptions,
+                    tunerOpts),
+                new ResourceBasedSlotSupplier(
+                    activityOptions == null
+                        ? new ResourceBasedSlotSupplierOptions()
+                        : activityOptions,
+                    tunerOpts),
+                new ResourceBasedSlotSupplier(
+                    localActivityOptions == null
+                        ? new ResourceBasedSlotSupplierOptions()
+                        : localActivityOptions,
+                    tunerOpts));
+        }
+
+        /// <summary>
+        /// Create a fixed-size tuner with the given slot capacities.
+        /// </summary>
+        /// <param name="numWorkflowTaskSlots">The number of available workflow task slots.</param>
+        /// <param name="numActivityTaskSlots">The number of available activity task slots.</param>
+        /// <param name="numLocalActivitySlots">The number of available local activity slots.</param>
+        /// <returns>The tuner.</returns>
+        public static WorkerTuner CreateFixed(
+            uint numWorkflowTaskSlots,
+            uint numActivityTaskSlots,
+            uint numLocalActivitySlots)
```

Comment:
While it makes more sense in Rust, it's not that common in .NET to use uints unless you _really_ need to (note our options don't really). Can check for negative at runtime if needed. Same type of thing as Go, most people pass around the common integer type and usually they aren't expected to convert for value-safety reasons.

Reply:
I'm not sure what would rise to the definition of "really need to" in the sense that things that shouldn't be negative shouldn't be negative and that's a fairly binary distinction.

It's annoying as hell that unsigned types exist in these languages but aren't used for whatever reason. Oh well though, I'll change it.

Reply:
Yeah, that is a bit annoying. I think the reasoning is more about the specific-bit-size integers and just carried over to here. For those, people avoid them unless they need specific binary sizes or are in a space-conscious place (e.g. embedded). And I think the general "use int for integers unless there is a good use case not to" was kinda the general accepted thing and applied to unsigned as well as specific-bit-size types. The only reason I bring up here is because in cases where uints or other things are used, .NET has a whole thing about unchecked conversions and it's a bit of a pain for devs using ints everywhere else.